TranscribeStreamingClientWrapper.java

```diff
@@ -0,0 +1,233 @@
+        if (requestStream != null) {
+            throw new IllegalStateException("Stream is already open");
+        }
+        try {
+            int sampleRate = 16_000; //default
+            if (inputFile != null) {
+                sampleRate = (int) AudioSystem.getAudioInputStream(inputFile).getFormat().getSampleRate();
+                requestStream = new AudioStreamPublisher(getStreamFromFile(inputFile));
+            } else {
+                requestStream = new AudioStreamPublisher(getStreamFromMic());
+            }
+            return client.startStreamTranscription(
+                    //Request parameters. Refer to API documentation for details.
+                    getRequest(sampleRate),
+                    //AudioEvent publisher containing "chunks" of audio data to transcribe
+                    requestStream,
+                    //Defines what to do with transcripts as they arrive from the service
+                    responseHandler);
+        } catch (LineUnavailableException | UnsupportedAudioFileException | IOException ex) {
+            CompletableFuture<Void> failedFuture = new CompletableFuture<>();
+            failedFuture.completeExceptionally(ex);
+            return failedFuture;
+        }
+    }
+
+    /**
+     * Stop in-progress transcription if there is one in progress by closing the request stream
+     */
+    public void stopTranscription() {
+        if (requestStream != null) {
+            try {
+                requestStream.inputStream.close();
+            } catch (IOException ex) {
+                System.out.println("Error stopping input stream: " + ex);
+            } finally {
+                requestStream = null;
+            }
+        }
+    }
+
+    /**
+     * Close clients and streams
+     */
+    public void close() {
+        try {
+            if (requestStream != null) {
+                requestStream.inputStream.close();
+            }
+        } catch (IOException ex) {
+            System.out.println("error closing in-progress microphone stream: " + ex);
+        } finally {
+            client.close();
+        }
+    }
+
+    /**
+     * Build an input stream from a microphone if one is present.
+     * @return InputStream containing streaming audio from system's microphone
+     * @throws LineUnavailableException When a microphone is not detected or isn't properly working
+     */
+    private static InputStream getStreamFromMic() throws LineUnavailableException {
+
+        // Signed PCM AudioFormat with 16kHz, 16 bit sample size, mono
+        int sampleRate = 16000;
+        AudioFormat format = new AudioFormat(sampleRate, 16, 1, true, false);
+        DataLine.Info info = new DataLine.Info(TargetDataLine.class, format);
+
+        if (!AudioSystem.isLineSupported(info)) {
+            System.out.println("Line not supported");
+            System.exit(0);
+        }
+
+        TargetDataLine line = (TargetDataLine) AudioSystem.getLine(info);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *line*. Currently, there are execution paths that do not contain closure statements, e.g., when exception is thrown by TargetDataLine.open. Either a) close *line* in a try-finally block or b) close the resource by declaring *line* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).